src/main/flight/mixer.c

```diff
@@ -347,14 +350,57 @@ static void applyFlipOverAfterCrashModeToMotors(void)
     }
 }
 
+#ifdef USE_RPM_LIMITER
+static void applyRpmLimiter(mixerRuntime_t *mixer)
+{
+    static float prevError = 0.0f;
+    static float i = 0.0f;
+    const float unsmoothedAverageRpm = getDshotAverageRpm();
+    const float averageRpm = pt1FilterApply(&mixer->averageRpmFilter, unsmoothedAverageRpm);
+    const float error = averageRpm - mixer->rpmLimiterRpmLimit;
+
+    // PID
+    const float p = error * mixer->rpmLimiterPGain;
+    const float d = (error - prevError) * mixer->rpmLimiterDGain; // rpmLimiterDGain already adjusted for looprate (see mixer_init.c)
+    i += error * mixer->rpmLimiterIGain;                          // rpmLimiterIGain already adjusted for looprate (see mixer_init.c)
+    i = MAX(0.0f, i);
+    float pidOutput = p + i + d;
+
+    // Throttle limit learning
+    if (error > -100.0f && rcCommand[THROTTLE] < rxConfig()->maxcheck) {
```

Comment:
> i think there is a little "bug" that when you drop your throttle down from 100% fast, it makes your `rpmLimiterThrottleScale` a few percent lower. i guess it needs a separate condition something like:
> 
> ```
> if ((throttle stick moving down) and (current rpm <= max allowed)) then
>     disable PID
> ```
I think the solution is to change the -100.0f magic number on line 370. I don't remember why it is set at 100, it should be 0.
```if (error > -100.0f && rcCommand[THROTTLE] < rxConfig()->maxcheck) {```
 to
```if (error > 0 && rcCommand[THROTTLE] < rxConfig()->maxcheck) {```

Reply:
any thoughts also why the initial guessing not working?

Reply:
My initial thought would be coming down from effective motor saturation, the average RPM observed doesn't drop all that much relative to the throttle input, so a couple percent change would happen from just that, and I wouldn't strictly consider that a problem as long as it's common to everybody flying it... realistically anytime I'm coming off sticks to plastic I'd prefer to have control authority overhead rather than the last 1% of thrust.
I'm still trying to understand the initial estimate being off

Reply:
@limonspb Because the guess is based on the **unloaded** RPM value. It guesses the RPM that the motor would spin at **without props**.

`unloaded_rpm = motor_kv * batteryVoltage`

This estimate will always be higher than the actual RPM with props.

Reply:
@TehllamaFPV What are you referring to?

In terms of the initial guessing, it’s worked in the past for me, I’ll check if it’s broken now for me

Reply:
@KarateBrot the problem is that it isn’t setting the expected throttle limit at all. When it does work, it works pretty good, and the throttle limit gets learned quickly

Reply:
@KarateBrot im talking about these lines of code, something off can be here:
```
void mixerResetRpmLimiter(void)
{
    const float maxExpectedRpm = MAX(1.0f, motorConfig()->kv * getBatteryVoltage() * 0.01f);
    mixerRuntime.rpmLimiterThrottleScale = constrainf(mixerRuntime.rpmLimiterRpmLimit / maxExpectedRpm, 0.0f, 1.0f);
}
```
this supposes to give a rough estimation of the current throttle limit right when you arm it.
But when i armed, in the log you can see its "100%". But it should be like 40-60%. (when running 18000 rpm limit 6S, 1960kv)

@Tdogb , i think the error might be here, instead
```
    const float maxExpectedRpm = MAX(1.0f, motorConfig()->kv * getBatteryVoltage() * 0.01f);
```
should be 
```
    const float maxExpectedRpm = MAX(1.0f, motorConfig()->kv * getBatteryVoltage());
```